### DISCUSSION TITLE
about js engine

### DISCUSSION THREAD

**Discussion by light4** [streamlink/streamlink#5020] [General]
Sorry, I just wonder why refuse to add a js engine like [js2py](https://github.com/PiotrDabkowski/Js2Py). With so many other [dependencies](https://github.com/streamlink/streamlink/blob/master/setup.cfg#L41).
[yt-dlp](https://github.com/yt-dlp/yt-dlp/) just create a small js engine in less than 1k lines.
https://github.com/yt-dlp/yt-dlp/blob/master/yt_dlp/jsinterp.py

#2945 #3068

Is there any chance to make [douyutv](https://github.com/light4/streamlink) works without a fork?

**Comment by gravyboat** [comment]
> Is there any chance to make [douyutv](https://github.com/light4/streamlink) works without a fork?

No. We're not willing to add a JS engine dependency for a single plugin.

**Reply by mkbloke** [reply]
> No. We're not willing to add a JS engine dependency for a single plugin.

To be fair, if there was a JS engine, it would probably be useful for more than one plugin eventually, but I can understand the reluctance to add more dependencies...

**Reply by gravyboat** [reply]
I'd want to see some interest in such plugins before doing so. We already have a lot of plugins that are either unused or barely used and since we don't collect any sort of usage statistics it's very difficult to justify adding another dependency and the maintenance burden that would bring.

**Comment by light4** [comment]
As a normal person, I only watch douyu.com stream sometimes.
I think it's normal for people just use this project to watch only a few websites.
So is the existence of other plugins in this project meanless? I doubt so.

**Comment by bastimeyer** [comment]
The yt-dlp implementation is very much tailored towards the needs of YouTube and its extractions for the data needed to calculate the stream/video URL signatures. That is only a tiny fraction of what a generic JS parser and interpreter do. Different sites do different stuff, meaning that implementing a generic JS engine requires having support for (almost) the entire ECMAScript 6 specification. Not even js2py does that, they do only support ECMAScript 5.1, which is very much dated. On top of that, lots of parts of the DOM also need to be implemented for compatibility of the JS code on websites, otherwise this is pointless. Given that all of that is evolving rapidly, even if there was a python project with good compatibility, there would be a high chance that this project would end up falling behind rather soon. Implementing all this in Streamlink is a totally ridiculous idea and doesn't belong here at all, only as a dedicated dependency. Apart from all that, a JS engine in Python is insanely slow and mostly only useful for extracing certain data, for example stuff that's hard to read with a regex or other means. Stuff that requires JS runtime calculations unfortunately is still better to be re-implemented in plugins. This is another approach that can be seen in other youtube projects related to streamlink/youtube-dl/yt-dlp.

**Reply by bastimeyer** [reply]
As an addition, what I forgot to mention is that a generic JS engine obviously allows execution of all possible JS code, which means there's always the halting problem, and running unknown/untrusted JS code from the perspective of Streamlink, even if sandboxed, is a terrible idea to begin with.
